File: 02_algorithm/programmers/Level3/(2020-KAKAO)기둥과 보 설치.py
def check_frame(answer):
    for frame in answer:
        x, y, a = frame
        if a == 0: # 기둥 설치 가능 조건
            if y == 0 or (x - 1, y, 1) in answer or (x, y, 1) in answer or (x, y - 1, 0) in answer:
                continue
            else:
                return False
        else: # 보 설치 가능 조건
            if (x, y - 1, 0) in answer or (x + 1, y - 1, 0) in answer or ((x - 1, y, 1) in answer and (x + 1, y, 1) in answer):
                continue
            else:
                return False
    return True

def solution(n, build_frame):
    answer = set()
    for frame in build_frame:
        x, y, a, b = frame
        if b == 1: # 설치하는 경우
            answer.add((x, y, a))
            if not check_frame(answer):
                answer.remove((x, y, a))
        else: # 삭제하는 경우
            answer.remove((x, y, a))
            if not check_frame(answer):
                answer.add((x, y, a))
    return sorted(answer)

print(solution(5, [
  [1, 0, 0, 1],
  [1, 1, 1, 1],
  [2, 1, 0, 1],
  [2, 2, 1, 1],
  [5, 0, 0, 1],
  [5, 1, 0, 1],
  [4, 2, 1, 1],
  [3, 2, 1, 1]
]))
print(solution(5, [
  [0, 0, 0, 1],
  [2, 0, 0, 1],
  [4, 0, 0, 1],
  [0, 1, 1, 1],
  [1, 1, 1, 1],
  [2, 1, 1, 1],
  [3, 1, 1, 1],
  [2, 0, 0, 0],
  [1, 1, 1, 0],
  [2, 2, 0, 1]
]))

File: 06_js/04_algorithm/programmers/Level3/(2020-KAKAO)기둥과 보 설치.js
const hasElementInSet = (set, element) => set.has(element.join('-'));

function solution(n, build_frame) {
  const checkConstructFrame = answer => {
    for (const frame of answer) {
      const [x, y, a] = frame.split('-').map(v => Number(v));
      if (a === 0) {
        if (y === 0 || hasElementInSet(answer, [x - 1, y, 1]) || hasElementInSet(answer, [x, y, 1]) || hasElementInSet(answer, [x, y - 1, 0])) {
          continue
        } else {
          return false
        }
      } else if (a === 1) {
        if (hasElementInSet(answer, [x, y - 1, 0]) || hasElementInSet(answer, [x + 1, y - 1, 0]) || (hasElementInSet(answer, [x - 1, y, 1]) && hasElementInSet(answer, [x + 1, y, 1]))) {
          continue
        } else {
          return false
        }
      }
    }
    return true
  }

  let answer = new Set();
  for (const frame of build_frame) {
    const [x, y, a, b] = frame;

    const toStringFrame = `${x}-${y}-${a}`;

    if (b === 1) {
      answer.add(toStringFrame);
      if (!checkConstructFrame(answer)) {
        answer.delete(toStringFrame);
      }
    } else {
      answer.delete(toStringFrame);
      if (!checkConstructFrame(answer)) {
        answer.add(toStringFrame);
      }
    }
  }
  return Array.from(answer).map(v => v.split('-').map(e => Number(e))).sort((a, b) => {
    if (a[0] === b[0]) {
      if (a[1] === b[1]) {
        return a[2] - b[2];
      }
      return a[1] - b[1];
    }
    return a[0] - b[0];
  })
}

console.log(solution(5, [
  [1, 0, 0, 1],
  [1, 1, 1, 1],
  [2, 1, 0, 1],
  [2, 2, 1, 1],
  [5, 0, 0, 1],
  [5, 1, 0, 1],
  [4, 2, 1, 1],
  [3, 2, 1, 1]
]))
console.log(solution(5, [
  [0, 0, 0, 1],
  [2, 0, 0, 1],
  [4, 0, 0, 1],
  [0, 1, 1, 1],
  [1, 1, 1, 1],
  [2, 1, 1, 1],
  [3, 1, 1, 1],
  [2, 0, 0, 0],
  [1, 1, 1, 0],
  [2, 2, 0, 1]
]))